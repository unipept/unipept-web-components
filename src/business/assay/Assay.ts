import { CountTable } from "@/business/counts/CountTable";
import Visitable from "@/logic/patterns/visitor/Visitable";

export default abstract class Assay implements Visitable<AssayVisitor> {
    protected constructor(
        public readonly changeListeners: ChangeListener<Assay>[],
        public readonly id: string,
        protected name?: string,
        protected date?: Date,
    ) {}

    public getId(): string {
        return this.id;
    }

    public getName() {
        return this.name;
    }

    public setName(name: string) {
        const oldName: string = this.name;
        this.name = name;
        this.onUpdate("name", oldName, name);
    }

    public getDate() {
        return this.date;
    }

    public setDate(date: Date) {
        const oldDate: Date = date;
        this.date = date;
        this.onUpdate("date", oldDate, date);
    }

    protected onUpdate(field: string, oldValue: any, newValue: any) {
        for (const changeListener of this.changeListeners) {
            changeListener.onChange(this, field, oldValue, newValue);
        }
    }

    abstract async accept(visitor: AssayVisitor): Promise<void>;
}

