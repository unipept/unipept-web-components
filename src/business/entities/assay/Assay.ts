import ChangeListener from "./../ChangeListener";
import AssayVisitor from "./AssayVisitor";
import Visitable from "./../../visitor/Visitable";

export default abstract class Assay implements Visitable<AssayVisitor> {
    private name: string = "";
    private date: Date = new Date();
    private changeListeners: ChangeListener<Assay>[] = [];

    protected constructor(
        public id: string
    ) {}

    public addChangeListener(listener: ChangeListener<Assay>): void {
        this.changeListeners.push(listener);
    }

    public getId(): string {
        return this.id;
    }

    public getName() {
        return this.name;
    }

    public setName(name: string) {
        const oldName: string = this.name;
        this.name = name;
        if (oldName !== name) {
            this.onUpdate("name", oldName, name);
        }
    }

    public getDate() {
        return this.date;
    }

    public setDate(date: Date) {
        const oldDate: Date = date;
        this.date = date;
        if (!oldDate || date.getTime() !== oldDate.getTime()) {
            this.onUpdate("date", oldDate, date);
        }
    }

    protected onUpdate(field: string, oldValue: any, newValue: any) {
        for (const changeListener of this.changeListeners) {
            changeListener.onChange(this, field, oldValue, newValue);
        }
    }

    abstract async accept(visitor: AssayVisitor): Promise<void>;
}

