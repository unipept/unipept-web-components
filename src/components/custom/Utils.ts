export default class Utils {
    /**
     * This method should be used when a specific URL should be opened in a new browser window. The method automatically
     * decides whether Electron or a default redirection should take place.
     * 
     * @param url The full url to which navigation should take place.
     */
    static openInBrowser(url: string): void {
        if (Utils.isElectron()) {
            const shell = require("electron").shell;
            shell.openExternal(url);
        } else {
            window.open(url);
        }
    }
    
    /**
     * Method that can be used to check if the app is being executed by Electron or something else.
     * 
     * @return True if the app is being executed in an Electron-powered environment.
     */
    static isElectron(): boolean {
        const userAgent = navigator.userAgent.toLowerCase();
        return userAgent.indexOf(" electron/") > -1;
    }
}
