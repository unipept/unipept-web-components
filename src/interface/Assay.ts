export default interface Assay {
    id: string
    name: string
    peptides: string[]
    amountOfPeptides: number
    createdAt: Date
};
