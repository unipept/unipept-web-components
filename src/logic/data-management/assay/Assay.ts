import DataRepository from "../../data-source/DataRepository";
import Visitable from "../../patterns/visitor/Visitable";
import AssayVisitor from "src/logic/data-management/assay/AssayVisitor";
import { StorageType } from "../StorageType";
import MPAConfig from "../MPAConfig";
import Entity from "./Entity";
import ChangeListener from "@/logic/data-management/ChangeListener";

export default abstract class Assay implements Visitable<AssayVisitor>, Entity<string> {
    protected id: string;
    protected name: string;
    protected date: Date;
    protected changeListener: ChangeListener;
    protected storageType: StorageType;

    protected _dataRepository: DataRepository;

    public progress: number = 0;

    constructor(changeListener?: ChangeListener, id?: string, storageType?: StorageType, name?: string, date?: Date) {
        this.id = id;
        this.name = name;
        this.date = date;
        this.changeListener = changeListener;
        this.storageType = storageType;
    }

    getId(): string {
        return this.id;
    }

    setId(id: string) {
        this.id = id;
        this.changeListener.onChange("id");
    }

    getName() {
        return this.name;
    }

    setName(name: string) {
        this.name = name;
        this.changeListener.onChange("name");
    }

    getDate() {
        return this.date;
    }

    setDate(date: Date) {
        this.date = date;
        this.changeListener.onChange("date");
    }

    getStorageType() {
        return this.storageType;
    }

    setStorageType(storageType: StorageType) {
        this.storageType = storageType;
    }

    get dataRepository() {
        return this._dataRepository;
    }

    getDateFormatted(): string {
        // @ts-ignore
        return this.date.getFullYear() + "/" + (this.date.getMonth() + 1).toString().padStart(2, "0") + "/" + this.date.getDate().toString().padStart(2, "0");
    }

    abstract async initDataRepository(mpaConfig: MPAConfig, baseUrl: string);
    abstract async accept(visitor: AssayVisitor): Promise<void>;
}
