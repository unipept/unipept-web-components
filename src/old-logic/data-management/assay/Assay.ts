import DataRepository from "../../data-source/DataRepository";
import Visitable from "../../patterns/visitor/Visitable";
import AssayVisitor from "./AssayVisitor";
import { StorageType } from "../StorageType";
import MPAConfig from "../MPAConfig";
import Entity from "./Entity";
import ChangeListener from "./../ChangeListener";

export default abstract class Assay implements Visitable<AssayVisitor>, Entity<string> {
    protected readonly id: string;
    protected name: string;
    protected date: Date;
    protected changeListeners: ChangeListener<Assay>[];
    protected storageType: StorageType;

    protected _dataRepository: DataRepository;

    public progress: number = 0;

    constructor(changeListeners: ChangeListener<Assay>[], id: string, storageType?: StorageType, name?: string, date?: Date) {
        this.id = id;
        this.name = name;
        this.date = date;
        this.changeListeners = changeListeners;
        this.storageType = storageType;
    }

    getId(): string {
        return this.id;
    }

    getName() {
        return this.name;
    }

    setName(name: string) {
        const oldName: string = this.name;
        this.name = name;
        this.onUpdate("name", oldName, name);
    }

    getDate() {
        return this.date;
    }

    setDate(date: Date) {
        const oldDate: Date = date;
        this.date = date;
        this.onUpdate("date", oldDate, date);
    }

    getStorageType() {
        return this.storageType;
    }

    setStorageType(storageType: StorageType) {
        this.storageType = storageType;
    }

    get dataRepository() {
        return this._dataRepository;
    }

    getDateFormatted(): string {
        // @ts-ignore
        return this.date.getFullYear() + "/" + (this.date.getMonth() + 1).toString().padStart(2, "0") + "/" + this.date.getDate().toString().padStart(2, "0");
    }

    public reset(): void {
        this.progress = 0;
        this._dataRepository = undefined;
    }

    protected onUpdate(field: string, oldValue: any, newValue: any) {
        for (const changeListener of this.changeListeners) {
            changeListener.onChange(this, field, oldValue, newValue);
        }
    }

    abstract async initDataRepository(mpaConfig: MPAConfig, baseUrl: string);
    abstract async accept(visitor: AssayVisitor): Promise<void>;
}
