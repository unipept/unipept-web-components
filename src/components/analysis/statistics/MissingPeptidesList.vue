<template>
    <div style="margin-top: 16px;">
        <div>
            Sorry, we didn't manage to find {{ missedPeptides.length }} of your
            peptides. You can BLAST them by clicking the links or copy them by using the button below.
        </div>

        <v-data-table :headers="headers" :items="missedPeptides" :items-per-page="10">
            <template v-slot:item.action="{ item }">
                <tooltip message="BLAST this peptide in an external window.">
                    <v-icon @click="openBlastUrl(item)">mdi-open-in-new</v-icon>
                </tooltip>
            </template>
        </v-data-table>

        <div
            style="margin-top: 8px; margin-bottom: 8px; display: flex; justify-content: center;"
            class="copy-button-container">
            <v-btn @click="copyToClipboard" color="primary">
                <v-icon left>
                    mdi-clipboard-text-outline
                </v-icon>
                Copy all to clipboard
            </v-btn>
        </div>
    </div>
</template>

<script lang="ts">
import Vue from "vue";
import { Component, Prop, Watch } from "vue-property-decorator";
import * as clipboard from "clipboard-polyfill";
import Utils from "./../../custom/Utils";
import Tooltip from "./../../custom/Tooltip.vue";
import { Peptide } from "@/business/ontology/raw/Peptide";

@Component({
    components: {
        Tooltip
    }
})
export default class MissingPeptidesList extends Vue {
    @Prop({ required: true })
    private missedPeptides: Peptide[] = [];

    private loading: boolean = true;

    private headers = [
        {
            text: "Peptide",
            align: "left",
            sortable: true,
            width: "80%",
            value: "value"
        },
        {
            text: "Actions",
            align: "center",
            sortable: false,
            width: "20%",
            value: "action"
        }
    ]

    private openBlastUrl(peptide: string) {
        const url: string = "http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&SET_SAVED_SEARCH=on" +
            "&USER_FORMAT_DEFAULTS=on&PAGE=Proteins&PROGRAM=blastp&QUERY=" + peptide + "&GAPCOSTS=11%201" +
            "&EQ_MENU=Enter%20organism%20name%20or%20id--completions%20will%20be%20suggested&DATABASE=nr" +
            "&BLAST_PROGRAMS=blastp&MAX_NUM_SEQ=100&SHORT_QUERY_ADJUST=on&EXPECT=10&WORD_SIZE=3" +
            "&MATRIX_NAME=BLOSUM62&COMPOSITION_BASED_STATISTICS=2&SHOW_OVERVIEW=on&SHOW_LINKOUT=on" +
            "&ALIGNMENT_VIEW=Pairwise&MASK_CHAR=2&MASK_COLOR=1&GET_SEQUENCE=on&NEW_VIEW=on&NUM_OVERVIEW=100" +
            "&DESCRIPTIONS=100&ALIGNMENTS=100&FORMAT_OBJECT=Alignment&FORMAT_TYPE=HTML&OLD_BLAST=false"
        Utils.openInBrowser(url);
    }

    private copyToClipboard() {
        clipboard.writeText(this.missedPeptides.join("\n"));
    }
}
</script>

<style>

</style>
