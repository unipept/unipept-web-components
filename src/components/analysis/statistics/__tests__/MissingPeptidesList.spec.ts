import { shallowMount, createLocalVue, mount } from "@vue/test-utils";
import MissingPeptidesList from "./../MissingPeptidesList.vue";
import Vue from "vue";
import Vuetify from "vuetify";
import Vuex from "vuex";
import Mock from "@/test/Mock";
import Assay from "@/logic/data-management/assay/Assay";
import flushPromises from "flush-promises";
import Utils from "./../../../custom/Utils";
import { contourDensity } from "d3";

jest.mock("./../../../custom/Utils");

Vue.use(Vuetify);
Vue.use(Vuex);

const localVue = createLocalVue();

// These peptides should be found by the tests
const foundPeptides: string[] = [
    "SGLAPFYSDK",
    "AAANESFGYNEDELVSSDLVGMR",
    "LAEEVLR",
    "YEVGTMLELPR",
    "TNTLLQSAFFK",
    "ANFEGECSEVGMYLAMAR",
    "MEVAVGDK",
    "LLDLGVLVGSGYHVNPK",
    "VYFLNFKPESSDEWK",
    "LGLVAVSR"
];

describe("MissingPeptidesList", () => {
    let vuetify;
    let getters;
    let store;

    beforeEach(() => {
        vuetify = new Vuetify();

        getters = {
            selectedTerm: () => "Organism",
            searchSettings: () => {
                return {
                    il: true,
                    dupes: true,
                    missed: false
                }
            }
        }

        store = new Vuex.Store({
            getters
        });
    });

    it("correctly renders all peptides", (done) => {
        let mock: Mock = new Mock();
        mock.mockInitializedAssay().then(async(assay: Assay) => {
            const wrapper = mount(MissingPeptidesList, {
                store,
                localVue,
                vuetify,
                propsData: {
                    dataset: assay 
                }
            });
    
            await sleep(1000);
            await flushPromises();

            // Test whether the calculated amount of missing peptides is correct.
            expect(wrapper.find("div > div").html()).toContain("we didn't manage to find 10 of your");
            expect(wrapper.find("div > div").html()).toMatchSnapshot();

            const dataTable = wrapper.find(".v-data-table").html();

            for (let peptide of foundPeptides) {
                expect(dataTable).toContain(peptide);
            }

            // Test if no more peptides are rendered.
            expect(wrapper.findAll("td.text-left").length).toEqual(10);

            done();
        });
    });

    it("correctly copies items to clipboard", (done) => {
        let mock: Mock = new Mock();
        mock.mockInitializedAssay().then(async(assay: Assay) => {
            const wrapper = mount(MissingPeptidesList, {
                store,
                localVue,
                vuetify,
                propsData: {
                    dataset: assay 
                }
            });
    
            await sleep(1000);
            await flushPromises();

            wrapper.find(".copy-button-container button").trigger("click");
            wrapper.vm.$nextTick(async() => {
                const content: string = await navigator.clipboard.readText();
                expect(content).toMatchSnapshot();
                done();
            })
        });
    });

    it("opens the correct URL for a peptide", (done) => {
        let mock: Mock = new Mock();
        mock.mockInitializedAssay().then(async(assay: Assay) => {
            const wrapper = mount(MissingPeptidesList, {
                store,
                localVue,
                vuetify,
                propsData: {
                    dataset: assay 
                }
            });
    
            await sleep(1000);
            await flushPromises();

            wrapper.find("td.text-center i").trigger("click");
            wrapper.vm.$nextTick(async() => {
                const expectedUrl: string = "http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&SET_SAVED_SEARCH=on" +
                "&USER_FORMAT_DEFAULTS=on&PAGE=Proteins&PROGRAM=blastp&QUERY=" + foundPeptides[0] + "&GAPCOSTS=11%201" + 
                "&EQ_MENU=Enter%20organism%20name%20or%20id--completions%20will%20be%20suggested&DATABASE=nr" +
                "&BLAST_PROGRAMS=blastp&MAX_NUM_SEQ=100&SHORT_QUERY_ADJUST=on&EXPECT=10&WORD_SIZE=3" + 
                "&MATRIX_NAME=BLOSUM62&COMPOSITION_BASED_STATISTICS=2&SHOW_OVERVIEW=on&SHOW_LINKOUT=on" + 
                "&ALIGNMENT_VIEW=Pairwise&MASK_CHAR=2&MASK_COLOR=1&GET_SEQUENCE=on&NEW_VIEW=on&NUM_OVERVIEW=100" + 
                "&DESCRIPTIONS=100&ALIGNMENTS=100&FORMAT_OBJECT=Alignment&FORMAT_TYPE=HTML&OLD_BLAST=false"

                expect(Utils.openInBrowser).toBeCalledWith(expectedUrl);
                done();
            });
        })
    });
})

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

