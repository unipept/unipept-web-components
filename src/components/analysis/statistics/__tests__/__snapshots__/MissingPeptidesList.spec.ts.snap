// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissingPeptidesList correctly copies items to clipboard 1`] = `
"SGLAPFYSDK
AAANESFGYNEDELVSSDLVGMR
LAEEVLR
YEVGTMLELPR
TNTLLQSAFFK
ANFEGECSEVGMYLAMAR
MEVAVGDK
LLDLGVLVGSGYHVNPK
VYFLNFKPESSDEWK
LGLVAVSR"
`;

exports[`MissingPeptidesList correctly renders all peptides 1`] = `
"<div>
        Sorry, we didn't manage to find 10 of your 
        peptides. You can BLAST them by clicking the links or copy them by using the button below.
    </div>"
`;
